

---

---
# Medidas Estatísticas

Este notebook apresenta cálculos de medidas estatísticas (de posição e de dispersão) utilizando diferentes funções (bibliotecas):

* Módulo de estatística do Python ([statistics](https://docs.python.org/3/library/statistics.html#module-statistics))
* Módulo de estatística da biblioteca Numpy ([statistics](https://numpy.org/doc/stable/reference/routines.statistics.html?highlight=statistics))
* Módulo de estatística da biblioteca Scipy ([stats](https://docs.scipy.org/doc/scipy/reference/stats.html?highlight=stats#module-scipy.stats))

Os cálculos são feitos em:

1. Array de dados
2. Dados agrupados em classes
3. Dataframes



---

---









---
## 1. Array de dados
---




In [1]:
import statistics # módulo de estatística do Python
import math # módulo de matemática do Python
import numpy as np # biblioteca numpy
from scipy import stats # módulo de estatística da biblioteca scipy

dados = np.array([24, 23, 22, 28, 35, 21, 23, 33, 34, 24, 21, 25, 36, 26, 22, 
                  30, 32, 25, 26, 33, 34, 21, 31, 25, 31, 26, 25, 35, 33, 31])

##Média

In [2]:
dados.sum() / len(dados)

27.833333333333332

In [3]:
dados.mean()

27.833333333333332

In [4]:
np.mean(dados)

27.833333333333332

In [5]:
statistics.mean(dados.astype(float))

27.833333333333332

##Mediana

In [6]:
np.median(dados)

26.0

In [7]:
statistics.median(dados)

26.0

##Moda

In [8]:
statistics.mode(dados)

25

In [9]:
stats.mode(dados)

ModeResult(mode=array([25]), count=array([4]))

##Quartis

In [10]:
np.quantile(dados,0.25) # 1o. quartil

24.0

In [11]:
np.quantile(dados,0.50) # 2o. quartil

26.0

In [12]:
np.quantile(dados,0.75) # 3o. quartil

32.75

##Percentis

In [27]:
np.percentile(dados,18) # percentil 18

23.0

## Variância

In [14]:
dados.var()

23.47222222222222

In [15]:
np.var(dados)

23.47222222222222

In [16]:
statistics.variance(dados)

24

##Desvio-padrão

In [17]:
dados.std()

4.844813951249544

In [18]:
np.std(dados)

4.844813951249544

In [19]:
statistics.stdev(dados)

4.898979485566356



---


##2. Dados agrupados em classes

---



Clona repositório de dados da disciplina hospedado no GitHUb

In [21]:
!git clone "https://github.com/malegopc/DSBD"

Cloning into 'DSBD'...
remote: Enumerating objects: 355, done.
remote: Counting objects: 100% (355/355), done.
remote: Compressing objects: 100% (332/332), done.
remote: Total 355 (delta 161), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (355/355), 6.14 MiB | 5.19 MiB/s, done.
Resolving deltas: 100% (161/161), done.


Leitura de tabela de distribuição de frequências de dados agrupados em classes

In [ ]:
import pandas as pd

tab = pd.read_csv('DSBD/Datasets/tab_dist_freq.csv')
# imprime as 10 primeiras linhas do dataframe
tab # mostra as 5 primeiras linhas e as 5 últimas do dataframe

Inf  Sup   Fi  Fac      fi    Xi
0    21   24  173  173  0.2253  22.5
1    24   27  127  300  0.1654  25.5
2    27   30   96  396  0.1250  28.5
3    30   33   61  457  0.0794  31.5
4    33   36   41  498  0.0534  34.5
5    36   39   51  549  0.0664  37.5
6    39   42   47  596  0.0612  40.5
7    42   45   39  635  0.0508  43.5
8    45   48   34  669  0.0443  46.5
9    48   51   18  687  0.0234  49.5
10   51   54   21  708  0.0273  52.5
11   54   57   13  721  0.0169  55.5
12   57   60   15  736  0.0195  58.5
13   60   63   11  747  0.0143  61.5
14   63   66    8  755  0.0104  64.5
15   66   69    8  763  0.0104  67.5
16   69   72    3  766  0.0039  70.5
17   72   75    1  767  0.0013  73.5
18   75   78    0  767  0.0000  76.5
19   78   81    1  768  0.0013  79.5

##Média de dados agrupados em classes

In [ ]:
# multiplica a frequência de cada classe pelo seu ponto médio (representante da classe)
tab['Fi.Xi'] = tab['Fi'] * tab['Xi'] 
tab

Inf  Sup   Fi  Fac      fi    Xi   Fi.Xi
0    21   24  173  173  0.2253  22.5  3892.5
1    24   27  127  300  0.1654  25.5  3238.5
2    27   30   96  396  0.1250  28.5  2736.0
3    30   33   61  457  0.0794  31.5  1921.5
4    33   36   41  498  0.0534  34.5  1414.5
5    36   39   51  549  0.0664  37.5  1912.5
6    39   42   47  596  0.0612  40.5  1903.5
7    42   45   39  635  0.0508  43.5  1696.5
8    45   48   34  669  0.0443  46.5  1581.0
9    48   51   18  687  0.0234  49.5   891.0
10   51   54   21  708  0.0273  52.5  1102.5
11   54   57   13  721  0.0169  55.5   721.5
12   57   60   15  736  0.0195  58.5   877.5
13   60   63   11  747  0.0143  61.5   676.5
14   63   66    8  755  0.0104  64.5   516.0
15   66   69    8  763  0.0104  67.5   540.0
16   69   72    3  766  0.0039  70.5   211.5
17   72   75    1  767  0.0013  73.5    73.5
18   75   78    0  767  0.0000  76.5     0.0
19   78   81    1  768  0.0013  79.5    79.5

In [ ]:
# soma a coluna Fi.Xi e divide pelo número elementos do conjunto de dados
n = tab['Fi'].sum() # número de elementos do conjunto de dados
x̄ = tab['Fi.Xi'].sum() / n
print('Média =', x̄)

Média = 33.8359375


## Moda em dados agrupados em classes

In [ ]:
# Encontra a classe modal
pos, = tab.index[tab['Fi'] == max(tab['Fi'])]
print('Classe Modal:',tab['Inf'][pos],'|-' ,tab['Sup'][pos])
print('Moda = ',tab['Xi'][pos])

Classe Modal: 21 |- 24
Moda =  22.5




---
##3. Dataframes
---





## Conjunto de dados

- Fonte: https://www.kaggle.com/uciml/pima-indians-diabetes-database
- Proprietários originais: National Institute of Diabetes and Digestive and Kidney Diseases
- Doadores da base de dados: Vincent Sigillito (vgs@aplcen.apl.jhu.edu), 
Research Center, RMI Group Leader, Applied Physics Laboratory, The Johns Hopkins University (1990)

Detalhes sobre o conjunto de dados:

1. Número de instâncias: 768

2. Número de atributos: 8 (mais classe target)

3. Atributos: (todos com valor numérico)
   1. Número de vezes que engravidou (*Pregnancies*)
   2. Concentração de glicose plasmática a 2 horas em um teste de tolerância à glicose oral (*Glucose*)
   3. Pressão arterial diastólica (mm Hg) (*BloodPressure*)
   4. Espessura da dobra da pele do tríceps (mm) (*SkinThickness*)
   5. Insulina sérica de 2 horas (mu U / ml) (*Insulin*)
   6. Índice de massa corporal (peso em kg / (altura em m) ^ 2) (*BMI*)
   7. Função de pedigree de diabetes (*DiabetesPedigreeFunction*)
   8. Idade (anos) (*Age*)
   
4. Variável target (classe): 0 ou 1 (valor de classe 1 é interpretado como "teste positivo para diabetes")

   Classe Valor Número de instâncias
   - 500 => 0
   - 268 => 1

 

## Leitura dos dados

Vamos começar importando o arquivo CSV "bruto" usando o Pandas.

In [22]:
import pandas as pd
# lê arquivo de dados, atribue NaN para dados faltantes
diabetes = pd.read_csv('DSBD/Datasets/Indians_Diabetes/diabetes.csv', na_values=['?'])
# imprime as 10 primeiras linhas do dataframe
diabetes # mostra as 5 primeiras linhas e as 5 últimas do dataframe

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

## Estatística descritiva 
Função describe() mostra um conjutonto de medidas estatísticas descritivas dos dados.

In [23]:
diabetes.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

## Média

In [24]:
diabetes['Age'].mean()

33.240885416666664

##Mediana

In [ ]:
diabetes['Age'].median()

29.0

##Quartil

In [25]:
diabetes['Age'].quantile(0.25)

24.0

##Desvio-padrão

In [26]:
diabetes['Age'].std()

11.76023154067868